### DISCUSSION TITLE
Add constraints on assignment problem_ Vehicle assignment

### DISCUSSION THREAD

**Discussion by vuphamngoctan1612** [google/or-tools#4138] [CP-SAT questions]
I have problem: How to choose best solution about cost for assign vehicles to routes. I want to use assignment problem algorithm in OR tool to resolve it

- List routes
  - Start_time
  - End_time
  - List locations of route

- List internal + external vehicles
  - Cost

I want to add constraints: time management https://colab.research.google.com/drive/1od_0IxyUKM5b4ITVbvsGo0pMmzbrXp90?usp=sharing

But it's not working. Why? and how to resolve it?

**Comment by lperron** [comment]
as usual, you write imperative code and not declarative code.

- You cannot use a variable as a boolean or integer value. variables are opaque objects.
- You cannot use python operators like or, and, min, max as they will silently create code that makes no sence.

I suggest:
  - switch to CP-SAT as (a) it is a better solver, and (b) it will not fail silently
  - look at ortools/sat/samples/assignment*.py
  - look at https://github.com/google/or-tools/blob/stable/ortools/sat/docs/channeling.md to understand how to write declarative code for tests

**Reply by vuphamngoctan1612** [reply]
![image](https://github.com/google/or-tools/assets/71437616/d115b827-afc7-4c44-a21e-0a668e147eb6)
When I run normal Assignment Problem, I have this result :
![image](https://github.com/google/or-tools/assets/71437616/d50e1962-e09d-4c27-af2c-37f2c4c42cb6)

I want to create a constraint: if car 1 is assigned to route A and B, solution is wrong because start - end of A and B are duplicate.
You have solution?

**Comment by lperron** [comment]
You mean overlap.

Just add all incompatibilities.

Le mer. 20 mars 2024, 07:11, Vũ Phạm Ngọc Tân ***@***.***> a
écrit :

**Reply by vuphamngoctan1612** [reply]
How to add all incompatibilities? I don't understand u

**Comment by lperron** [comment]
interval x overlap with interval y
for all workers w: assign[w][x] + assign[w][y] <= 1

Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le mer. 20 mars 2024 à 10:14, Vũ Phạm Ngọc Tân ***@***.***> a
écrit :

**Reply by vuphamngoctan1612** [reply]
In my result, worker 0 is assigned to 2 tasks: 0 and 1
But task 0 start at 8 and end at 10; 
task 1 start at 8.5 and end at 11;
So this solution is wrong, and algorithm have to find other solution

**Comment by lperron** [comment]
exactly my point
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le mer. 20 mars 2024 à 11:32, Vũ Phạm Ngọc Tân ***@***.***> a
écrit :
